extends Node

@export var mob_scene: PackedScene
var score

# Receive hit() signal from Player node and name as game_over() func
func game_over() -> void:
	$ScoreTimer.stop()
	$MobTimer.stop()

func new_game() -> void:
	score = 0
	$Player.start($StartPosition.position)
	$StartTimer.start()
	
# Gives Player a few second, then actually starts the game
func _on_start_timer_timeout() -> void:
	$MobTimer.start()
	$ScoreTimer.start()
	
func _on_score_timer_timeout() -> void:
	score += 1
	
func _on_mob_timer_timeout() -> void:
	# Creates a mob
	var mob = mob_scene.instantiate()
	
	# MobPath is Path2D node. MobSpawnLocation is PathFollow2D node.
	var mob_spawn_location = $MobPath/MobSpawnLocation
	
	# Choose a random location on the path to spawn mob
	mob_spawn_location.progress_ratio = randf()
	mob.position = mob_spawn_location.position
	
	# Set the mob's direction perpendicular to the path direction.
	var direction = mob_spawn_location.rotation + PI / 2
	
	# Add some randomness to the direction.
	direction += randf_range(-PI / 4, PI / 4)
	mob.rotation = direction
	
	# Choose the velocity for the mob.
	var velocity = Vector2(randf_range(150.0, 250.0), 0.0)
	mob.linear_velocity = velocity.rotated(direction)

	# Spawn the mob by adding it to the Main scene.
	add_child(mob)	

func _ready():
	new_game()
